File: examples/05_traits/main.cpp
// ============================================
// 05 — Traits & Generics (C++)
// ============================================
// Kompilieren: g++ -std=c++17 main.cpp -o main && ./main

#include <iostream>
#include <vector>
#include <memory>
#include <cmath>

// --- Beispiel 1: Pure virtual class = Rust trait ---
class Area {
public:
    virtual double area() const = 0; // pure virtual — wie Rust trait
    virtual std::string describe() const {
        return "Form mit Fläche " + std::to_string(area());
    }
    virtual ~Area() = default;
};

struct Circle : public Area {
    double radius;
    Circle(double r) : radius(r) {}
    double area() const override { return M_PI * radius * radius; }
};

struct Rectangle : public Area {
    double width, height;
    Rectangle(double w, double h) : width(w), height(h) {}
    double area() const override { return width * height; }
};

// --- Beispiel 2: Dynamischer Dispatch (vtable) ---
// Entspricht Rust: fn print_area(shape: &dyn Area)
void print_area_dynamic(const Area& shape) {
    std::cout << "[dynamic] " << shape.describe() << std::endl;
}

// --- Beispiel 3: Templates = statischer Dispatch ---
// Entspricht Rust: fn print_area(shape: &impl Area)
// Problem: Kein explizites Interface — Fehler erst bei Instanziierung
template<typename T>
void print_area_static(const T& shape) {
    std::cout << "[static]  " << shape.describe() << std::endl;
    // Was T sein muss? Unklar aus der Signatur!
    // In Rust: fn print_area<T: Area>(shape: &T) — explizit
}

// --- Beispiel 4: C++20 Concepts — expliziter wie Rust Traits ---
// (Rust hatte das schon immer durch Trait Bounds)
#if __cplusplus >= 202002L
#include <concepts>
template<typename T>
concept HasArea = requires(T t) {
    { t.area() } -> std::convertible_to<double>;
};

template<HasArea T>
void print_area_concept(const T& shape) {
    std::cout << "[concept] " << shape.area() << std::endl;
}
#endif

int main() {
    Circle    c{3.0};
    Rectangle r{4.0, 5.0};

    print_area_static(c);
    print_area_static(r);

    // Heterogene Collection — braucht Pointer (unique_ptr)
    // Entspricht Rust: Vec<Box<dyn Area>>
    std::vector<std::unique_ptr<Area>> shapes;
    shapes.push_back(std::make_unique<Circle>(1.0));
    shapes.push_back(std::make_unique<Rectangle>(2.0, 3.0));
    shapes.push_back(std::make_unique<Circle>(5.0));

    for (const auto& s : shapes) {
        print_area_dynamic(*s);
    }

    return 0;
}

File: examples/05_traits/main.rs
// ============================================
// 05 — Traits & Generics
// ============================================
// Analogie C++:
//   class Shape { virtual double area() = 0; }  → trait Area
//   template<typename T>                         → fn foo<T: Trait>
//
// Rust:
//   impl Trait  → statischer Dispatch (Zero-Cost, wie C++ Templates)
//   dyn Trait   → dynamischer Dispatch (vtable, wie C++ virtual)

use std::fmt;

// --- Trait Definition (wie pure virtual class in C++) ---
trait Area {
    fn area(&self) -> f64;
    fn describe(&self) -> String {  // Default-Implementierung möglich
        format!("Form mit Fläche {:.2}", self.area())
    }
}

struct Circle { radius: f64 }
struct Rectangle { width: f64, height: f64 }

impl Area for Circle {
    fn area(&self) -> f64 {
        std::f64::consts::PI * self.radius * self.radius
    }
}

impl Area for Rectangle {
    fn area(&self) -> f64 {
        self.width * self.height
    }
}

// --- Statischer Dispatch (wie C++ Templates — Zero-Cost) ---
fn print_area_static(shape: &impl Area) {
    println!("[static]  {}", shape.describe());
}

// --- Dynamischer Dispatch (wie C++ virtual — vtable) ---
fn print_area_dynamic(shape: &dyn Area) {
    println!("[dynamic] {}", shape.describe());
}

// --- Generics mit Bounds (explizit, klare Fehlermeldungen) ---
fn largest<T: PartialOrd>(list: &[T]) -> &T {
    let mut largest = &list[0];
    for item in list {
        if item > largest { largest = item; }
    }
    largest
}

// --- Mehrere Trait Bounds (wie C++20 Concepts) ---
fn print_and_compare<T: fmt::Display + PartialOrd>(a: T, b: T) {
    if a > b {
        println!("{} ist größer als {}", a, b);
    } else {
        println!("{} ist kleiner/gleich {}", a, b);
    }
}

fn main() {
    let c = Circle { radius: 3.0 };
    let r = Rectangle { width: 4.0, height: 5.0 };

    print_area_static(&c);
    print_area_static(&r);

    // Heterogene Collection mit dyn Trait (wie vector<unique_ptr<Shape>>)
    let shapes: Vec<Box<dyn Area>> = vec![
        Box::new(Circle { radius: 1.0 }),
        Box::new(Rectangle { width: 2.0, height: 3.0 }),
        Box::new(Circle { radius: 5.0 }),
    ];
    for shape in &shapes {
        print_area_dynamic(shape.as_ref());
    }

    // Generics
    let numbers = vec![34, 50, 25, 100, 65];
    println!("Größte Zahl: {}", largest(&numbers));

    print_and_compare(3.14, 2.71);
}
